### ISSUE TITLE
Why is the AngleSharp dependency not at 1.0?

### ISSUE THREAD

**Issue by mganss** [mganss/HtmlSanitizer#491]
TL;DR HtmlSanitizer depends on AngleSharp.Css whose stable version is not at 1.0 yet and is incompatible with AngleSharp 1.0. If you want to use AngleSharp 1.0, use the latest prerelease version of HtmlSanitizer.

See also #87, #451 for in-depth discussions of this topic.

**Comment by adnan-kamili** [comment]
I am unable to upgrade the PreMailer.Net 2.6.0 -> AngleSharp (>= 1.1.0) 
because HtmlSanitizer 8.0.865  depends on AngleSharp (= 0.17.1).

**Comment by mganss** [comment]
@adnan-kamili Use the prerelease version of HtmlSanitizer (currently [8.1.866-beta](https://www.nuget.org/packages/HtmlSanitizer/8.1.866-beta))

**Comment by jamesperrin** [comment]
Has this topic of targeting the latest version of AngleSharp been revisited?

**Comment by mganss** [comment]
@jamesperrin AngleSharp.Css is still not at 1.0 yet.

**Comment by sandcastle** [comment]
> It seems like work on [AngleSharp.CSS](https://github.com/AngleSharp/AngleSharp.Css) is decreasing. Almost no work for 17 months, with last _code_ PR completed in February 2024 😕

There are commits as late as Aug 025
https://github.com/AngleSharp/AngleSharp.Css/commit/aec713b40e3455b625b8c925b0e6b6cb916852e4

**Comment by IrvinDominin** [comment]
Hi any news about the update to the latest AngleSharp versions?

**Comment by tiesont** [comment]
> Hi any news about the update to the latest AngleSharp versions?

Until you see a 1.0 release of AngleSharp.Css, you're not likely to see @mganss unpinning the AngleSharp version. The linked issues already explain why.

**Comment by IrvinDominin** [comment]
> > Hi any news about the update to the latest AngleSharp versions?
> 
> Until you see a 1.0 release of AngleSharp.Css, you're not likely to see [@mganss](https://github.com/mganss) unpinning the AngleSharp version. The linked issues already explain why.

Thank you @tiesont hoping will be released soon to moving forward, we have some dependecies issue that cannot be resolved actually and we can't use the beta release

**Comment by pawlos** [comment]
is there any update of using any newer version of `AngleSharp` in this project? Keeping the current one exposes to a potential `NullReferenceException` from a short, untrusted input. The bug is present in `AngleSharp` (0.17.1.0) but has been fixed in some version between 0.17.1.0 and the latest. @mganss Shall I report this as a separate issue?

**Comment by mganss** [comment]
@pawlos Do you mean report the AngleSharp bug at the AngleSharp repo? Sure, go ahead.

The blocker still is the 1.0 release of AngleSharp.Css. Ideally, someone would help out @FlorianRappl to clear that hurdle. Not sure what his policy is on AI use, but perhaps that could help to reduce the required effort to reach 1.0. See https://github.com/AngleSharp/AngleSharp.Css/pull/157 on what needs to be done.

**Comment by tiesont** [comment]
@mganss Not sure what the feature parity is like, but the HtmlAgilityPack project (https://html-agility-pack.net/) was taken over by ZZZ Projects, so there is at least one alternative angle bracket processor out there. Like AngleSharp, though, the CSS parsing is a bit lacking (I think you have to add Fizzler/Hazz (https://github.com/atifaziz/Hazz) for any real CSS support).

Just thought I'd mention it - been on my to-do list for a while to see if HtmlSanitizer could be made to work with HAP without breaking changes.

**Comment by pawlos** [comment]
> [@pawlos](https://github.com/pawlos) Do you mean report the AngleSharp bug at the AngleSharp repo? Sure, go ahead.

No. I was thinking to add it here but I guess it doesn't make sense except being aware of it. The bug is in AngleSharp but it was fixed there. It can be triggered via HtmlSanitizer due to this old dependency on AngleSharp.

**Comment by jafin** [comment]
> It seems like work on [AngleSharp.CSS](https://github.com/AngleSharp/AngleSharp.Css) is decreasing. Almost no work for 17 months, with last _code_ PR completed in February 2024 😕

Pr's were submitted in 2026.  So there is _has been work_ on it.  Need to find out from the maintainer (@FlorianRappl) what their release cadence is.

**Comment by omaristalis** [comment]
Anglesharp 0.17.1 now has a moderate vulnerability (https://github.com/advisories/GHSA-pgww-w46g-26qg)

Do you know when a non-prerelease version of HtmlSanitizer will be available that references a newer version?

**Comment by NVedsted** [comment]
I had hoped we could perhaps get a backported fix to 0.x of AngleSharp but it seems quite settled that it is considered EOL. As an alternative, we are considering if we can mitigate the vulnerability by disallowing `<annotation-xml>` or  perhaps even `<math>` altogether.

EDIT: Seems like these tags are not allowed by default, so perhaps not as big a deal as we feared.

**Comment by FlorianRappl** [comment]
We discussed this quite in depth if this is even worth a vulnerability as AngleSharp does not execute anything or do anything harmful here. However, under certain circumstances (and HtmlSanatizer might trigger one of these) a trust could be established where the generated output is actually leading to a vulnerability (the construction involving the mentioned tags to inject scripts that would otherwise - i.e., if nothing is done, not appear in the "dangerous" positions). For 99% of the users there is 0 chance that this has any impact - nevertheless, we wanted to make sure that also the 1% is properly secured - even with that kind of a low chance of actually triggering it.

The backport fix is unfortunately not feasible - we'd like to move forward as this is the most efficient use of resources.

Hope that helps and thanks for reaching out @NVedsted.

**Comment by mganss** [comment]
AngleSharp.Css 1.0 stable release will come within the next two weeks 🙌🏻 https://github.com/AngleSharp/AngleSharp.Css/pull/157#issuecomment-5036340419

**Comment by dsparkplug** [comment]
Or even sooner: https://github.com/AngleSharp/AngleSharp.Css/pull/157/changes/2bfee1564e876e08f923b0f3e4cdce5972eefb70 🎉

**Comment by mjolivet-lucca** [comment]
a dependabot pull request is available : https://github.com/mganss/HtmlSanitizer/pull/620

**Comment by OlegoO** [comment]
https://github.com/AngleSharp/AngleSharp.Css/releases/tag/v1.0.0

**Comment by mganss** [comment]
I've just released HtmlSanitizer 9.1.973 which has AngleSharp 1.6.0 and AngleSharp.Css 1.0.0 as dependencies. 

A huge thanks to @FlorianRappl for the awesome job he's done on AngleSharp for many years. To everyone who's using AngleSharp and/or HtmlSanitizer: Please consider donating to the AngleSharp project. Here's more: https://github.com/AngleSharp/AngleSharp/issues/1163

So glad I can finally close this one! 🎉
